controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -786,7 +790,20 @@ func (r *BareMetalHostReconciler) actionPreparing(prov provisioner.Provisioner,
 		prepareData.ActualRAIDConfig = nil
 		dirty = true
 	}
+
+	// Use settings in hostFirmwareSettings if available
+	hfs, err := r.getValidHostFirmwareSettings(info)
```

Comment:
There's a couple of points here that could deserve some attention/discussion.

The first one is related to the validation of the HFS settings: I can appreciate that re-applying it synchronously here may eliminate some issues, but in order to promote a better separation of the responsibilities between the controllers it seems more an operation that should belong to the HFS controller. In particular, its reconcile loop may take care of validating the current specs as soon as they are updated, and flag the HFS resource as valid/not-valid. In that way the BMH controller will have only to wait for a valid HFS resource here, before proceeding with the `Preparing` state, and it will become just a mere HFS reader (with the only responsibility to assemble the required data for the clean step). The HFS flag could be a simple boolean or, maybe better for future extensions, a proper state.

The second point it's still about the synchronization between the two controllers. Since they are independent, there could be a potential race (the effective probabilities are very low), where the HFS controller didn't yet fill correctly the HFS resource while the BMH controller arrives at this point and wants to go on. An explicit synchronization mechanism based on flag or status field could eliminate also this problem at all.

Reply:
Sure, I moved the validation to the HFS controller and added a bool that can be read by the BMH controller. The flag is set after the Spec is compared to the Schema.